src/replication.c

```diff
@@ -0,0 +1,1067 @@
+	replica->id = spec->replica_count;
+	spec->replica_count++;
+	spec->consistency_count = spec->replica_count/2 +1;
+	TAILQ_INSERT_TAIL(&spec->rq, replica, r_next);
+	if(spec->replica_count == 1)
+		pthread_cond_signal(&spec->rq_cond);
+	MTX_UNLOCK(&spec->rq_mtx);
+	rc = pthread_create(&replica_sender_thread, NULL, &replica_sender,
+			(void *)replica);
+	if (rc != 0) {
+		ISTGT_ERRLOG("pthread_create(replicator_thread) failed\n");
+		return NULL;
+	}	
+	zvol_io_hdr_t *rio;
+	rio = (zvol_io_hdr_t *)malloc(sizeof(zvol_io_hdr_t));
+	rio->opcode = ZVOL_OPCODE_HANDSHAKE;
+	rio->io_seq = 0;
+	rio->offset = 0;
+	char *data = malloc(strlen("vol1")+1);
+	rio->len    = strlen("vol1");
+	strncpy(data, "vol1", strlen("vol1"));
+	
+	write(replica->iofd, rio, sizeof(zvol_io_hdr_t));
+	write(replica->iofd, data, strlen("vol1"));
+	free(rio);
+	rio = NULL;
+	spec->ready = true;
+	return replica;
+}
+
+int
+zvol_handshake(char *volname, char *replicaip, int replica_port) {
+	int rc, iofd;
+	spec_t *spec;
+	REPLICA_LOG("VOLNAME = %s\n", volname);
+	MTX_LOCK(&specq_mtx);
+	TAILQ_FOREACH(spec, &spec_q, spec_next) {
+		if(strcmp(volname, volname) == 0)  {
+			if((iofd = cstor_ops.conn_connect(replicaip, replica_port)) < 0) {
+				REPLICA_ERRLOG("conn_connect() failed errno:%d\n", errno);
+				exit(EXIT_FAILURE);
+			}
+			rc = make_socket_non_blocking(iofd);
+			if (rc == -1) {
+				REPLICA_ERRLOG("make_socket_non_blocking() failed errno:%d\n", errno);
+				exit(EXIT_FAILURE);
+
+			}
+			create_replica_entry(spec, iofd, replicaip, replica_port);
+			break;
+		}
+	}
+	MTX_UNLOCK(&specq_mtx);
+	return 0;
+}
+
+int
+accept_mgmt_conns(int epfd, int sfd) {
+	struct epoll_event event;
+	int rc, rcount=0;
+	spec_t *spec;
+	int mgmtfd[MAXREPLICA];
+	char *buf = malloc(BUFSIZE);
+	zvol_op_code_t mgmt_opcode;
+	int mgmt_fd, data_len = 0;
+
+	while (1) {
+		struct sockaddr saddr;
+		socklen_t slen;
+		char hbuf[NI_MAXHOST], sbuf[NI_MAXSERV];
+		slen = sizeof(saddr);
+		mgmt_fd = accept(sfd, &saddr, &slen);
+		if (mgmt_fd == -1) {
+			if((errno == EAGAIN) || (errno == EWOULDBLOCK)) {
+				break;
+			} else {
+				REPLICA_ERRLOG("accept() failed, errrno:%d", errno);
+				break;
+			}
+		}
+
+		rc = getnameinfo(&saddr, slen,
+				hbuf, sizeof(hbuf),
+				sbuf, sizeof(sbuf),
+				NI_NUMERICHOST | NI_NUMERICSERV);
+		if (rc == 0) {
+			mgmtfd[rcount] = mgmt_fd;
+			rcount++;
+			REPLICA_LOG("Accepted connection on descriptor %d "
+					"(host=%s, port=%s)\n", mgmt_fd, hbuf, sbuf);
+		}
+		rc = make_socket_non_blocking(mgmt_fd);
+		if (rc == -1) {
+			REPLICA_ERRLOG("make_socket_non_blocking() failed, errno:%d", errno);
+			return -1;
+		}
+		event.data.fd = mgmt_fd;
+		event.events = EPOLLIN | EPOLLET;
+		rc = epoll_ctl(epfd, EPOLL_CTL_ADD, mgmt_fd, &event);
+		if(rc == -1) {
+			REPLICA_ERRLOG("epoll_ctl() failed, errno:%d", errno);
+			return -1;
+		}
+		MTX_LOCK(&specq_mtx);
+		TAILQ_FOREACH(spec, &spec_q, spec_next) {
+			data_len += snprintf(buf, BUFSIZE, "%s", "vol1") + 1;
```

Comment:
this need to be made configurable, and its at multiple places.

Reply:
done in latest commit